p32/src/bank.rs

```diff
@@ -0,0 +1,248 @@
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Bank {
+    pub name: String,
+    pub users: Vec<User>,
+    pub credit_interest: u64, // in basis points (0.01%)
+    pub debit_interest: u64,  // in basis points (0.01%)
+}
+
+#[derive(Debug)]
+pub enum TransferError {
+    UserNotFound(String),
+    InsufficientFunds(String),
+    CreditLimitExceeded(String),
+}
+
+impl fmt::Display for TransferError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            TransferError::UserNotFound(name) => write!(f, "User {} not found", name),
+            TransferError::InsufficientFunds(name) => {
+                write!(f, "User {} has insufficient funds", name)
+            }
+            TransferError::CreditLimitExceeded(name) => {
+                write!(f, "User {} would exceed credit line", name)
+            }
+        }
+    }
+}
+
+impl Error for TransferError {}
+
+impl User {
+    pub fn new(name: String, credit_line: u64, balance: i64) -> Self {
+        Self {
+            name,
+            credit_line,
+            balance,
+        }
+    }
+}
+
+impl Bank {
+    pub fn new(name: String, credit_interest: u64, debit_interest: u64) -> Self {
+        Self {
+            name,
+            users: Vec::new(),
+            credit_interest,
+            debit_interest,
+        }
+    }
+
+    pub fn add_user(&mut self, user: User) {
+        self.users.push(user);
+    }
+
+    pub fn calc_balance(&self) -> (u64, u64) {
+        let mut liabilities = 0;
+        let mut assets = 0;
+
+        for user in &self.users {
+            if user.balance > 0 {
+                liabilities += user.balance as u64;
+            } else {
+                assets += (-user.balance) as u64;
+            }
+        }
+
+        (liabilities, assets)
+    }
+
+    pub fn transfer_funds(
+        &mut self,
+        from_name: &str,
+        to_name: &str,
+        amount: u64,
+    ) -> Result<(), TransferError> {
+        let amount_i64 = amount as i64;
+
+        // Find user indices first
+        let from_idx = self
+            .users
+            .iter()
+            .position(|u| u.name == from_name)
+            .ok_or(TransferError::UserNotFound(from_name.to_string()))?;
+
+        let to_idx = self
+            .users
+            .iter()
+            .position(|u| u.name == to_name)
+            .ok_or(TransferError::UserNotFound(to_name.to_string()))?;
+
+        // Check if transfer is possible
+        if self.users[from_idx].balance < amount_i64 {
+            return Err(TransferError::InsufficientFunds(from_name.to_string()));
+        }
+
+        if (self.users[from_idx].balance - amount_i64).unsigned_abs()
+            > self.users[from_idx].credit_line
+        {
+            return Err(TransferError::CreditLimitExceeded(from_name.to_string()));
+        }
+
+        if (self.users[to_idx].balance + amount_i64).unsigned_abs() > self.users[to_idx].credit_line
```

Comment:
You don't need to check recipient since we increase his balance, so he can not go below credit limit in this operation.